### PR TITLE
[Merged by Bors] - feat(Data/Setoid/Partition): equivalence classes and quotient

### PR DESCRIPTION
**Pull request by chrisflav**
Adds the equivalence between the type of equivalence classes of an equivalence relation and its quotient.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by eric-wieser — COMMENTED** [review]
Can you add a lemma that says what `classesEquivQuotient (Quotient.mk r a)` is equal to?

File: Mathlib/Data/Setoid/Partition.lean [hunk 1]
```diff
@@ -197,6 +197,28 @@ theorem sUnion_classes (r : Setoid α) : ⋃₀ r.classes = Set.univ :=
   Set.eq_univ_of_forall fun x => Set.mem_sUnion.2 ⟨{ y | r.Rel y x }, ⟨x, rfl⟩, Setoid.refl _⟩
 #align setoid.sUnion_classes Setoid.sUnion_classes
 
+/-- The equivalence between the type of equivalence classes of an equivalence relation and its
+quotient. -/
+noncomputable def classes_equiv_quotient (r : Setoid α) : Quotient r ≃ Setoid.classes r := by
```
[the remark this answers is not in the document]
Comment (chrisflav) [reply Mathlib/Data/Setoid/Partition.lean]: Fixed.

**Comment by chrisflav** [comment]
> Can you add a lemma that says what `classesEquivQuotient (Quotient.mk r a)` is equal to?

I did, I was not sure whether to add this as elements of `Set α` or as elements of the subtype `Setoid.classes r`, so I added both. If you want me to remove one of them, please say so.

File: Mathlib/Data/Setoid/Partition.lean [hunk 2]
```diff
@@ -197,6 +197,36 @@ theorem sUnion_classes (r : Setoid α) : ⋃₀ r.classes = Set.univ :=
...
+    intro ⟨c, a, hc⟩
+    exact ⟨a, Subtype.ext hc.symm⟩
+
+lemma classesEquivQuotient_mk_eq (r : Setoid α) (a : α) :
+    classesEquivQuotient r (Quotient.mk r a) = ⟨{ x | r.Rel x a }, Setoid.mem_classes r a⟩ :=
+  rfl
+
+lemma classesEquivQuotient_mk_coe_eq (r : Setoid α) (a : α) :
+    (classesEquivQuotient r (Quotient.mk r a) : Set α) = { x | r.Rel x a } :=
+  Subtype.ext_iff_val.mp (classesEquivQuotient_mk_eq r a)
```
Comment (eric-wieser) [inline Mathlib/Data/Setoid/Partition.lean:228]:
```suggestion
  rfl
```

File: Mathlib/Data/Setoid/Partition.lean [hunk 3]
```diff
@@ -197,6 +197,36 @@ theorem sUnion_classes (r : Setoid α) : ⋃₀ r.classes = Set.univ :=
...
+    exact Setoid.refl a
+  · rw [Quot.surjective_lift]
+    intro ⟨c, a, hc⟩
+    exact ⟨a, Subtype.ext hc.symm⟩
+
+lemma classesEquivQuotient_mk_eq (r : Setoid α) (a : α) :
+    classesEquivQuotient r (Quotient.mk r a) = ⟨{ x | r.Rel x a }, Setoid.mem_classes r a⟩ :=
+  rfl
+
+lemma classesEquivQuotient_mk_coe_eq (r : Setoid α) (a : α) :
```
Comment (eric-wieser) [inline Mathlib/Data/Setoid/Partition.lean:226]:
```suggestion
@[simp] lemma classesEquivQuotient_mk_coe_eq (r : Setoid α) (a : α) :
```

File: Mathlib/Data/Setoid/Partition.lean (hunk 2 above)
Comment (chrisflav) [reply Mathlib/Data/Setoid/Partition.lean:228]: This does not seem to work, that's why I added the `Subtype.ext_iff_val.mp`.

File: Mathlib/Data/Setoid/Partition.lean (hunk 3 above)
Comment (chrisflav) [reply Mathlib/Data/Setoid/Partition.lean:226]: Are you happy with the name? I am honestly a bit lost with the naming conventions.

File: Mathlib/Data/Setoid/Partition.lean [hunk 4]
```diff
@@ -197,6 +197,33 @@ theorem sUnion_classes (r : Setoid α) : ⋃₀ r.classes = Set.univ :=
   Set.eq_univ_of_forall fun x => Set.mem_sUnion.2 ⟨{ y | r.Rel y x }, ⟨x, rfl⟩, Setoid.refl _⟩
 #align setoid.sUnion_classes Setoid.sUnion_classes
 
+/-- The equivalence between the type of equivalence classes of an equivalence relation and its
+quotient. -/
+noncomputable def classesEquivQuotient (r : Setoid α) : Quotient r ≃ Setoid.classes r := by
```
Comment (j-loreaux) [inline Mathlib/Data/Setoid/Partition.lean:202]:
```suggestion
noncomputable def quotientEquivClasses (r : Setoid α) : Quotient r ≃ Setoid.classes r := by
```

File: Mathlib/Data/Setoid/Partition.lean [hunk 5]
```diff
@@ -197,6 +197,33 @@ theorem sUnion_classes (r : Setoid α) : ⋃₀ r.classes = Set.univ :=
...
+    apply Quotient.sound
+    show a ∈ { x | Setoid.r x b }
+    rw [←h_eq]
+    exact Setoid.refl a
+  · rw [Quot.surjective_lift]
+    intro ⟨c, a, hc⟩
+    exact ⟨a, Subtype.ext hc.symm⟩
+
+@[simp]
+lemma classesEquivQuotient_mk_eq (r : Setoid α) (a : α) :
```
Comment (j-loreaux) [inline Mathlib/Data/Setoid/Partition.lean:223]:
```suggestion
lemma quotientEquivClasses_mk_eq (r : Setoid α) (a : α) :
```

**Review by j-loreaux — COMMENTED** [review]
these are just named backwards.

File: Mathlib/Data/Setoid/Partition.lean (hunk 4 above)
Comment (chrisflav) [reply Mathlib/Data/Setoid/Partition.lean]: Fixed in the new commit, I also adapted the doc string to reflect the order.

File: Mathlib/Data/Setoid/Partition.lean (hunk 5 above)
Comment (chrisflav) [reply Mathlib/Data/Setoid/Partition.lean]: Fixed in the new commit.

**Comment by j-loreaux** [comment]
Thanks!

bors merge

**Pull request closed without merging**